modules/jolt_physics/objects/jolt_soft_body_3d.cpp

```diff
@@ -424,6 +424,134 @@ bool JoltSoftBody3D::is_sleeping() const {
+
+	ERR_FAIL_NULL(shared);
+	ERR_FAIL_INDEX(p_index, (int)shared->mesh_to_physics.size());
+	const size_t physics_index = (size_t)shared->mesh_to_physics[p_index];
+
+	if (pinned_vertices.has(physics_index)) {
+		return;
+	}
+
+	const float last_step = space->get_last_step();
+	if (unlikely(last_step == 0.0f)) {
+		return;
+	}
+	wake_up();
+
+	JoltWritableBody3D body = space->write_body(jolt_id);
+	ERR_FAIL_COND(body.is_invalid());
+
+	JPH::SoftBodyMotionProperties &motion_properties = static_cast<JPH::SoftBodyMotionProperties &>(*body->GetMotionPropertiesUnchecked());
+
+	JPH::Array<JPH::SoftBodyVertex> &physics_vertices = motion_properties.GetVertices();
+	JPH::SoftBodyVertex &physics_vertex = physics_vertices[physics_index];
+
+	const JPH::Vec3 impulse = to_jolt(p_impulse);
+	physics_vertex.mVelocity += impulse * physics_vertex.mInvMass;
+}
+
+void JoltSoftBody3D::apply_node_force(int p_index, const Vector3 &p_force) {
+	ERR_FAIL_COND_MSG(!in_space(), vformat("Failed to apply force for '%s'. Doing so without a physics space is not supported when using Jolt Physics. If this relates to a node, try adding the node to a scene tree first.", to_string()));
+
+	ERR_FAIL_NULL(shared);
+	ERR_FAIL_INDEX(p_index, (int)shared->mesh_to_physics.size());
+	const size_t physics_index = (size_t)shared->mesh_to_physics[p_index];
+
+	if (pinned_vertices.has(physics_index)) {
+		return;
+	}
+
+	const float last_step = space->get_last_step();
+	if (unlikely(last_step == 0.0f)) {
+		return;
+	}
+	wake_up();
+
+	JoltWritableBody3D body = space->write_body(jolt_id);
+	ERR_FAIL_COND(body.is_invalid());
+
+	JPH::SoftBodyMotionProperties &motion_properties = static_cast<JPH::SoftBodyMotionProperties &>(*body->GetMotionPropertiesUnchecked());
+
+	JPH::Array<JPH::SoftBodyVertex> &physics_vertices = motion_properties.GetVertices();
+	JPH::SoftBodyVertex &physics_vertex = physics_vertices[physics_index];
+
+	const JPH::Vec3 force = to_jolt(p_force);
+	physics_vertex.mForce += force;
+}
+
+void JoltSoftBody3D::apply_central_impulse(const Vector3 &p_impulse) {
+	ERR_FAIL_COND_MSG(!in_space(), vformat("Failed to apply central impulse for '%s'. Doing so without a physics space is not supported when using Jolt Physics. If this relates to a node, try adding the node to a scene tree first.", to_string()));
+
+	ERR_FAIL_NULL(shared);
+
+	const float last_step = space->get_last_step();
+	if (unlikely(last_step == 0.0f)) {
+		return;
+	}
+	wake_up();
+
+	JoltWritableBody3D body = space->write_body(jolt_id);
+	ERR_FAIL_COND(body.is_invalid());
+
+	JPH::SoftBodyMotionProperties &motion_properties = static_cast<JPH::SoftBodyMotionProperties &>(*body->GetMotionPropertiesUnchecked());
+
+	JPH::Array<JPH::SoftBodyVertex> &physics_vertices = motion_properties.GetVertices();
+
+	const int mesh_vertex_count = shared->mesh_to_physics.size();
+
+	const JPH::Vec3 impulse = to_jolt(p_impulse) / mesh_vertex_count;
```

Comment:
It's very debatable if dividing by mesh_vertex_count is the right thing to do or what applying a central impulse on a soft body even means. 

E.g.
* What if half of the vertices are pinned? This means that half of the impulse is applied to pinned vertices which is lost since they won't move, resulting in applying only half the impulse to the body
* What if the mass of the vertices is not uniform (I don't think this is currently possible)? Then some particles will accelerate less than others causing the body to stretch.

Maybe you should calculate the total delta velocity for the entire body by calculating p_impulse / total_mass_of_all_vertices and then applying that delta velocity to all vertices. Take care not to change the velocity of pinned vertices.

Reply:
> * What if half of the vertices are pinned? This means that half of the impulse is applied to pinned vertices which is lost since they won't move, resulting in applying only half the impulse to the body

So, like, the more vertices pinned, the stronger the impulse on the unpinned vertices. That felt kind of unrealistic, isn't it?

> What if the mass of the vertices is not uniform (I don't think this is currently possible)? Then some particles will accelerate less than others causing the body to stretch.

Wouldn't it just get pulled back together to prevent it from stretching too far.

Reply:
@jrouwe 
Do you still want me to try your idea? Because, I notice that you fixed the "apply central force" on [jolt](https://github.com/jrouwe/JoltPhysics/commit/7850b05a97d2079fc52e538507c843026a555ef3), and all you did is just divide it by the vertex count. So it should be the same for the impulse. If the implementation right now is fine, can you continue on reviewing it? Thanks.

Reply:
Sorry, I did the review last week but I appear to have forgotten to 'finish' the review